### PR TITLE
[RooFit] Replace RooArgSet and RooDataSet mempools with single templated mempool

### PR DESCRIPTION
**Pull request by hageboeck**
This fixes ROOT-9571 and other crashes when RooPlots reach the end of main.

One may wonder whether mempools are a good idea. The answer is probably no, but disabling the mempool breaks the caching of normalisation results in RooFit. This relies on RooArgSet pointers being unique, i.e. OS-managed memory allocation doesn't work.

With the proposed implementation, there is just one (templated) mempool for both Roo(Arg|Data)Set. Possible future modifications (e.g. locks, thread-local arenas) need to be done only in one place and are way easier now. The mempool takes care of not deleting (in worst case leaking) any arenas with live objects at program end. This brings us much closer to solving all problems with global statics in RooFit.

### CODE REVIEW

**Comment by hageboeck** [comment]
2 TODOs (mostly note to self before merging):
* Increase size of arenas a bit
* Add static-destruction-order test to roottest

**Comment by dpiparo** [comment]
@hageboeck , nice improvement. A comment only partially related to this PR which could be merged immediately (this is a bug fix and can go in 6.16). Being forward looking: I agree that custom allocators are not a very good idea, especially when thinking to sophisticated stuff such as thread safety.
If the pointer of the RooArgSet is used as unique identifyier, why don't we add a unique id to the RooArgSet instances and rely on that one rather than their pointers?

File: roofit/roofitcore/inc/RooArgSet.h [hunk 1]
```diff
@@ -21,12 +21,14 @@
 class RooArgList ;
 
 
-#define USEMEMPOOL
+#define USEMEMPOOLFORARGSET
```
Comment (dpiparo) [inline roofit/roofitcore/inc/RooArgSet.h]: Is there a case where the guard is not set?

Comment (hageboeck) [reply roofit/roofitcore/inc/RooArgSet.h]: I don't see what you mean. It's not a guard, but a switch. However, the switch (old name) was switching both memory pools for RooArgSet or RooDataSet, or only one - depending on the order of inclusion.
Therefore, I changed the names of both switches such that each only switches on/off for one of the two sets.

**Pull request merged**